account_invoice_ubl_mass_download/models/account_invoice.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2021 Coop IT Easy (https://coopiteasy.be)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import base64
+import io
+import logging
+import zipfile
+
+from odoo import api, models
+
+_logger = logging.getLogger(__name__)
+
+ATTACHMENT_TMP_NAME = "temporary-invoice-ubl-zip-file"
+
+
+class AccountInvoice(models.Model):
+
+    _inherit = "account.invoice"
+
+    @api.multi
+    def zip_ubl_xml_and_pdf(self):
+        """
+        Create a zip file with XML UBL and PDF for each invoice.
+        """
+        with io.BytesIO() as buffer:
+            with zipfile.ZipFile(
+                buffer, mode="w", compression=zipfile.ZIP_DEFLATED
+            ) as ziparc:
+                for invoice in self:
+                    assert invoice.type in ("out_invoice", "out_refund")
+                    assert invoice.state in ("open", "paid")
```

Comment:
Less a comment than a question: I've always been under the impression that relying on asserts in Python for business logic is bad practice, because asserts can be disabled using `python -O`, but I've seen asserts here and there in Odoo code. Is this normal in Odoo, or are the asserts still bad practice?

If bad practice, these should probably be `if-raise`.

Reply:
I agree that using `assert` here is not fancy and that a proper `raise` with a proper explanation message will be better.

I just use it in the same way it was used in other function of the parent module. So that I ensure the same behaviour that in the rest of the parent module.